download/download.go

```diff
@@ -0,0 +1,163 @@
+package download
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/alexandrasp/taskcluster-cli/extpoints"
+	"github.com/taskcluster/httpbackoff"
+	"github.com/taskcluster/taskcluster-client-go"
+	"github.com/taskcluster/taskcluster-client-go/queue"
+)
+
+//"github.com/taskcluster/httpbackoff"
+
+func init() {
+	extpoints.Register("download", download{})
+}
+
+type download struct{}
+
+func (download) ConfigOptions() map[string]extpoints.ConfigOption {
+	return nil
+}
+
+func (download) Summary() string {
+	return "Download an artifact"
+}
+
+func usageDownload() string {
+	return `Usage:
+			taskcluster download [options]
+			Options:
+			<taskId> [<runId>] <artifact>
+			`
+}
+
+func (download) Usage() string {
+	return usageDownload()
+}
+
+func (download) Execute(context extpoints.Context) bool {
+	command := context.Arguments["download"].(string)
+	taskId := context.Arguments["<taskId>"].(string)
+	runId := context.Arguments["<runId>"].(string)
+	artifact := context.Arguments["<artifact>"].(string)
+
+	provider := extpoints.CommandProviders()[command]
+	if provider == nil {
+		log.Panicf("Unknown command %s", command)
+
+	} else {
+
+		permaCred := &tcclient.Credentials{
+			ClientID:    os.Getenv("TASKCLUSTER_CLIENT_ID"),
+			AccessToken: os.Getenv("TASKCLUSTER_ACCESS_TOKEN"),
+		}
+
+		userQueue := queue.New(permaCred)
+
+		if runId != "" {
+			//get a artifact with runId parameter
+			url_artifact, err := userQueue.GetArtifact_SignedURL(taskId, runId, artifact, time.Second*10)
+			if err != nil {
+				log.Panicf("Exception thrown signing URL \n%s", err)
+			} else {
+
+				url_artifact := EnforceHttpsUrl(url_artifact.String())
+				fmt.Printf("NEW URL %s\n", url_artifact)
+				response, attempts, err := getAnArtifact(url_artifact)
+
+				if err != nil {
+					log.Panicf("Exception thrown download an artifact \n%s", err)
+				} else {
+					fmt.Printf("Number of attempts: %d\n", attempts)
+					_, length, out := checkContentLength(response)
+					log.Printf("ContentLength %d with %s", length, out)
+				}
+			}
+		}
+		if runId == "" {
+			//get latest artifact without rundId parameter
+			url_artifact, err := userQueue.GetLatestArtifact_SignedURL(taskId, artifact, time.Second*10)
+			if err != nil {
+				log.Panicf("Exception thrown signing URL \n%s", err)
```

Comment:
much of this code is duplicated, but it only really differs in `GetArtifact_SignedURL` and `GetLatestArtifact_SignedURL` calls.